### DISCUSSION TITLE
v23 能不能支持支持触摸板多指手势，尤其是三指左右、上下切换应用

### DISCUSSION THREAD

**Discussion by pmzgit** [linuxdeepin/developer-center#4506] [Features Request & Ideas | 特性请求 & 头脑风暴]
- 从v20开始就关注这个问题，也在论坛讨论过，社区产品经理已经记到需求里了，我也提交到过bug提交入口，但是到现在还是没能实现。提到github，希望能尽快实现。

- mac和win10+，对于触摸板手势配置支持的很全， **产品阶段完全不用时间，交互逻辑参考win10的完全可以(这个不涉及抄袭、模仿的问题，触摸板手势而已，而且统一手势，本来就是应该这样的)**  。即使不提供配置入口，参考win10默认的多指手势，也已经符合主流操作逻辑，对于我们这些平时不用鼠标(除非有精确文字编辑需求的，再用鼠标，当然其实纯用触摸板也没啥问题)的人来说，**切换应用总是alt+tab，一个一个切换到指定应用，太低效了** (肯定有人说，你可以触摸板上，单指移动到任务栏的指定应用，切换应用，是可以，但不如三指滑动高效，这是毋庸置疑的，不用讨论)
- 我把win11的触摸板配置入口，发过来，以方便没win环境的开发人员参考：
<img width="900" alt="1685195900413" src="https://github.com/linuxdeepin/developer-center/assets/11833302/362ecda5-4516-465b-ad08-45daf906b490">

- 我用manjro的时候，发现，触摸板手势支持也是用的gesture方案，但是这个方案不够好，好多手势不支持，后来我用https://github.com/iberianpig/fusuma 方案配置manjaro的kde，发现是可行的，三指左右滑动切换，但是上下滑动我没时间配置(因为当时正在调研arch系，当然，最后还是回到了deepin，比kde用着舒服，也更好看，养老了确定了！！！不过使用distrobox(archlinux容器)配合deepin使用) 。所以我觉得fusuma这个方案是可行的，当然我预计需要dde的修改，所以我估计即使自己尝试配置，也不会成功
- 我参考论坛，找到配置触摸板的入口（sudo nvim /usr/share/dde-daemon/gesture.json），但是，发现dde内置的手势（less /usr/share/dde-daemon/gesture_builtin.md），虽然支持三指左右滑动切换应用，但是没有驻留应用列表的能力(就是：按下alt+tab健，然后松开tab健，不松开alt健，此时的状态是显示所有应用列表)，而且更不用说三指上下滑动，切换第二排的应用列表。还是那意思，使用win10 或 win11，体验一下多指操作的魅力。
- 我把当时自己调研的配置发一下(当然没配置完全，没完全实现三指滑动的交互逻辑)
<img width="641" alt="1685197130364" src="https://github.com/linuxdeepin/developer-center/assets/11833302/fbc37fe1-66fd-4947-b288-f7519b2379ea">

![image](https://github.com/linuxdeepin/developer-center/assets/11833302/e28dd5fd-4d03-4160-8fec-deb95013a745)

- 作为开发人员，我一直致力于研究，使用**系统提供的快捷键+触摸板多指手势，来代替低效的鼠标，提升pc操作效率（至于为啥鼠标低效，我觉得不用质疑吧，你手来回离开返回键盘区域，其实是不止低效，而且最重要会打断你大脑的思路(准确来讲，打断心流状态)）**
- 为此，我自己总结了高效操作的一些逻辑，如下图，关于触摸板应该支持的手势，及手势映射的操作，都写明了
<img width="1140" alt="1685194602302" src="https://github.com/linuxdeepin/developer-center/assets/11833302/81fb2f21-60dc-4fda-993e-9d8e9805a990">

- 我觉得其他手势可以都不支持，毕竟单指、两指，其实目前系统都支持了，四指其实不用管，毕竟切换桌面需求不大，但是三指滑动，是目前最急需的硬需求。但是详细来讲交互操作逻辑来说，实现的时候一定要注意，**当三指滑动开始后，停住不动，应用列表应该持续显示，并且一定是两排显示。让用户看到，需要切换的目标应用在哪，再上下、左右滑动到指定应用。** 我在win11上截图，类似如下图：

<img width="1238" alt="1685195038852" src="https://github.com/linuxdeepin/developer-center/assets/11833302/d0a3904a-9238-4649-bd23-46585c91952f">

- 其实我猜测深度开发组，不愿意切换触摸板实现方案的原因，**触摸板的驱动兼容性**，gesture方案能保证大多数触摸板最起码能用？fusuma方案可能导致触摸板驱动不支持？但是我又感觉，图形栈的交互流程应该是一致的吧，触摸板驱动接收滑动、点击等事件，并将事件或中断发送到内核，libinput(还是什么库)接收事件、中断，xdotool(等类似库)操作桌面图形等。
- 我想如果dde支持这个三指手势，在论坛宣传一下，我相信，肯定有很多人会感觉非常需要，尤其是那些开发人员，普通用户也会觉得惊喜好用。我觉得真的是个硬需求，毕竟都23年了，触摸板三指手势支持，应该是必备特性了吧
- 最后，如果官方不打算当前做，希望了解触摸板手势的deepin开发人员，能给个手动实现方案，我自己来测试，并尝试配置

**Comment by pmzgit** [comment]
mint21.2 已经支持多指手势
https://www.oschina.net/news/245580/linux-mint-touchpad-gestures-support
mint选择的开源方案：
https://github.com/JoseExposito/touchegg

**Comment by pmzgit** [comment]
https://github.com/JoseExposito/touchegg#keyboard-shortcut-send_keys

https://github.com/iberianpig/fusuma/wiki/3-fingers-Alt-Tab-Switcher(Windows-Style)

**Comment by BLumia** [comment]
23 在发正式版之前肯定不会做新需求了，目前精力在解决 23 现存的问题。至于触控板手势要不要在 23 正式版后考虑做的事情，cc @meiyixiang 

> 最后，如果官方不打算当前做，希望了解触摸板手势的deepin开发人员，能给个手动实现方案，我自己来测试，并尝试配置

无论 DDE 现有维护团队打不打算做，作为开发者都可以考虑自己调研并尝试实现。我看你其实自己已经调研过比较多的方案了，不妨直接试试看。只是我对触控板手势的了解也不多所以可能目前帮不了太多忙，不妨来 https://wiki.deepin.org/%E4%BA%A4%E6%B5%81%E6%96%B9%E5%BC%8F 里的开发者讨论群里一起讨论，兴许有感兴趣的小伙伴一起搞...

**Comment by meiyixiang** [comment]
目前触碰版有一套自己的基础操作能力支持，但没有提供自定义和修改的能力。
短时间内不会考虑输出此问题的相关需求，目前23版本，还有太多基础能力带去规划和实现。
![image](https://github.com/linuxdeepin/developer-center/assets/75475994/78f837b9-66cc-40f3-bfd5-2ff76b849ce0)

**Comment by pmzgit** [comment]
行吧，主要手头没有合适设备，公司配备的台式机没有触摸板。个人电脑因vpn等问题，暂时无法切换deepin，等回头有时间我再尝试吧
